cmd/soroban-rpc/internal/events/events.go

```diff
@@ -0,0 +1,276 @@
+package events
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/xdr"
+)
+
+// Cursor represents the position of a Soroban event.
+// Soroban events are sorted in ascending order by
+// ledger sequence, transaction index, operation index,
+// and event index.
+type Cursor struct {
+	// Ledger is the sequence of the ledger which emitted the event.
+	Ledger uint32
+	// Tx is the index of the transaction within the ledger which emitted the event.
+	Tx uint32
+	// Op is the index of the operation within the transaction which emitted the event.
+	Op uint32
+	// Event is the index of the event within in the operation which emitted the event.
+	Event uint32
+}
+
+func cmp(a, b uint32) int {
+	if a < b {
+		return -1
+	}
+	if a > b {
+		return 1
+	}
+	return 0
+}
+
+// Cmp compares two cursors.
+// 0 is returned if the c is equal to other.
+// 1 is returned if c is greater than other.
+// -1 is returned if c is less than other.
+func (c Cursor) Cmp(other Cursor) int {
+	if c.Ledger == other.Ledger {
+		if c.Tx == other.Tx {
+			if c.Op == other.Op {
+				return cmp(c.Event, other.Event)
+			}
+			return cmp(c.Op, other.Op)
+		}
+		return cmp(c.Tx, other.Tx)
+	}
+	return cmp(c.Ledger, other.Ledger)
+}
+
+var (
+	// MinCursor is the smallest possible cursor
+	MinCursor = Cursor{}
+	// MaxCursor is the largest possible cursor
+	MaxCursor = Cursor{
+		Ledger: math.MaxUint32,
+		Tx:     math.MaxUint32,
+		Op:     math.MaxUint32,
+		Event:  math.MaxUint32,
+	}
+)
+
+type bucket struct {
+	ledgerSeq uint32
+	events    []event
+}
+
+type event struct {
+	contents   xdr.ContractEvent
+	txIndex    uint32
+	opIndex    uint32
+	eventIndex uint32
+}
+
+func (e event) cursor(ledger uint32) Cursor {
+	return Cursor{
+		Ledger: ledger,
+		Tx:     e.txIndex,
+		Op:     e.opIndex,
+		Event:  e.eventIndex,
+	}
+}
+
+// MemoryStore is an in-memory store of soroban events.
+type MemoryStore struct {
+	lock sync.RWMutex
+	// buckets is a circular buffer where each cell represents
+	// all events occurring within a specific ledger.
+	buckets []bucket
+	// length is equal to the number of ledgers contained within
+	// the circular buffer.
+	length uint32
+	// start is the index of the head in the circular buffer.
+	start uint32
+}
+
+// NewMemoryStore creates a new MemoryStore.
+// The retention window is in units of ledgers.
+// All events occurring in the following ledger range
+// [ latestLedger - retentionWindow, latestLedger ]
+// will be included in the MemoryStore. If the MemoryStore
+// is full, any events from new ledgers will evict
+// older entries outside the retention window.
+func NewMemoryStore(retentionWindow uint32) (*MemoryStore, error) {
+	if retentionWindow == 0 {
+		return nil, fmt.Errorf("retention window must be positive")
+	}
+	return &MemoryStore{
+		buckets: make([]bucket, retentionWindow),
```

Comment:
by having `length` member in `MemoryStore`, you're duplicating the inner member. I can't find any place where this is an issue in the codebase, but in general, it would be safer to use `make([]bucket, 0, retentionWindow)` and use `len(buckets)` to find the length of the buckets and `cap(buckers)` to check the max size.